File: problems/leetcode/0900/981/981.cpp
// 981. Time Based Key-Value Store
// https://leetcode.com/problems/time-based-key-value-store/description/

using namespace std;

#include <string>
#include <unordered_map>
#include <vector>

class TimeMap {
public:
  unordered_map<string, vector<pair<int, string>>> store;

  TimeMap() {}

  int upperBoundForStore(string key, int timestamp) {
    if (store[key].empty()) {
      return -1;
    }

    if (timestamp >= store[key].back().first) {
      return store[key].size();
    }

    if (timestamp < store[key][0].first) {
      return -1;
    }

    int start = 0, end = store[key].size() - 1;

    while (start < end) {
      int mid = (start + end) / 2;
      if (store[key][mid].first <= timestamp) {
        start = mid + 1;
      } else {
        end = mid;
      }
    }

    return end;
  }

  void set(string key, string value, int timestamp) {
    store[key].push_back({timestamp, value});
  }

  string get(string key, int timestamp) {
    int index = upperBoundForStore(key, timestamp);

    if (index <= 0) {
      return "";
    }

    return store[key][index - 1].second;
  }
};

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap* obj = new TimeMap();
 * obj->set(key,value,timestamp);
 * string param_2 = obj->get(key,timestamp);
 */

File: problems/leetcode/0900/981/981.py
# 981. Time Based Key-Value Store
# https://leetcode.com/problems/time-based-key-value-store/description/

import bisect
from collections import defaultdict

class TimeMap:

    def __init__(self):
        self.keyToValues = defaultdict(list[str]) # (str, List[str])
        self.keyToTimestamps = defaultdict(list[int]) # (str, List[int])

    def set(self, key: str, value: str, timestamp: int) -> None:
        self.keyToValues[key].append(value)
        self.keyToTimestamps[key].append(timestamp)

    def get(self, key: str, timestamp: int) -> str:
        if not self.keyToTimestamps[key]:
            return ""

        index = bisect.bisect_right(self.keyToTimestamps[key], timestamp) - 1
        return self.keyToValues[key][index] if index >= 0 else ""


# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)
